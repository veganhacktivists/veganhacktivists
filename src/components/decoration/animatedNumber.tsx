'use client';

import { useSpring, animated } from '@react-spring/web';
import classNames from 'classnames';
import React, { useMemo, useState } from 'react';
import { Waypoint } from 'react-waypoint';

import useReduceMotion from '../../hooks/useReduceMotion';

const sizeRegex = /([a-z]{1,3}:)?text-[0-9][^\s]+/g;

const AnimatedNumber: React.FC<{
  number: number;
  approx?: boolean;
  prefix?: string;
  suffix?: string;
  className?: string;
}> = ({ number, className = '', approx = false, prefix, suffix }) => {
  const [onView, setOnView] = useState<boolean>(false);

  const prefersReducedMotion = useReduceMotion();

  const { number: interpolatedNumber } = useSpring({
    from: { number: 0 },
    to: { number },
    config: { duration: prefersReducedMotion ? 0 : 500 },
    cancel: !onView,
  });

  const [sizeClasses, classesWithoutSize] = useMemo(() => {
    const sizeClasses = Array.from(className.matchAll(sizeRegex)).map(
      ([result]) => result,
    );

    const classesWithoutSize = className.replace(sizeRegex, '');
    return [sizeClasses, classesWithoutSize];
  }, [className]);

  const classes = classNames(
    'font-bold font-mono',
    classesWithoutSize,
    sizeClasses.length ? sizeClasses : 'text-7xl md:text-8xl',
  );

  return (
    <>
      <Waypoint
        onEnter={() => {
          setOnView(true);
        }}
      />
      <span className={classes} aria-label={`${number}`}>
        {prefix}
        <animated.span>
          {interpolatedNumber.to((x) =>
            Math.floor(x)
              .toString()
              .replace(/\B(?=(\d{3})+(?!\d))/g, ','),
          )}
        </animated.span>
        {suffix}
        {approx && <>~</>}
      </span>
    </>
  );
};

export default AnimatedNumber;
