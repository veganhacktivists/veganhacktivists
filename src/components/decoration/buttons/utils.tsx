import classNames from 'classnames';
import React from 'react';
import getThemeColor from '../../../lib/helpers/theme';

interface FillBackgroundProps extends React.PropsWithChildren {
  base: string;
  fill: string;
  disabled?: boolean;
}

const fillStyle: (from: string, to: string) => React.CSSProperties = (
  from,
  to
) => {
  return {
    backgroundImage: `linear-gradient(to right, ${getThemeColor(
      to
    )},  50%, ${getThemeColor(from)} 50%)`,
    backgroundSize: '250% 100%',
    backgroundPosition: 'right',
  };
};

export const FillBackground: React.FC<FillBackgroundProps> = ({
  base,
  fill,
  disabled = false,
  children,
}) => {
  const childrenWithProps = React.Children.map(children, (child) => {
    if (!React.isValidElement(child)) return;
    const props = child.props;
    return React.cloneElement(child, {
      className: classNames(
        props.className,
        'transition-background-position duration-[400ms] ease-linear',
        {
          'hover:!bg-left group-hover:!bg-left': !disabled,
        }
      ),
      style: { ...props.style, ...(disabled ? {} : fillStyle(base, fill)) },
    });
  });

  return <>{childrenWithProps}</>;
};
