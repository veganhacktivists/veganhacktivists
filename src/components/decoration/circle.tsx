interface CircleProps {
  color?: string;
  opacity?: number;
  thickness?: string;
  xAlign?: 'left' | 'right';
  yAlign?: 'top' | 'bottom';
  radiusZoom?: number;
}

const Circle: React.FC<CircleProps> = ({
  color = 'white',
  opacity = 0.5,
  thickness = '20px',
  yAlign = 'top',
  xAlign = 'left',
  radiusZoom = 1,
  ...trimmedProps
}) => {
  const radius = (100 * radiusZoom) / 3;

  const xPos = xAlign == 'left' ? 'left-0' : 'right-0';
  const yPos = yAlign == 'top' ? 'top-0' : 'bottom-0';

  let xTransform = 'translate-x-1/4';
  let yTransform = 'translate-y-2/4 lg:translate-y-3/4';

  if (xAlign == 'left') xTransform = '-translate-x-1/4';
  if (yAlign == 'top') yTransform = '-translate-y-2/4 lg:-translate-y-3/4';

  return (
    <svg
      width={radius * 2 + 'vw'}
      height={radius * 2 + 'vw'}
      className={`text-${color} overflow-visible absolute ${yPos} ${xPos} transform ${xTransform} ${yTransform}
    `}
    >
      <circle
        {...trimmedProps}
        r={radius + 'vw'}
        strokeWidth={thickness}
        fillOpacity={0}
        opacity={opacity}
        className="stroke-current"
        cx={radius + 'vw'}
        cy={radius + 'vw'}
      />
    </svg>
  );
};

export default Circle;
