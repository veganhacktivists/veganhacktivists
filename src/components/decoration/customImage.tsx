/* eslint-disable react/forbid-elements */
/* eslint-disable jsx-a11y/alt-text */
import classNames from 'classnames';
import Image from 'next/image';

import type { ImageProps } from 'next/image';

interface CustomImageProps extends ImageProps {
  useNextImage?: boolean;
}

const CustomImage: React.FC<CustomImageProps> = ({ priority, ...props }) => {
  if (props.useNextImage === false) {
    return (
      // required for the image to redered to a string for transaltions
      // @ts-expect-error ---
      // eslint-disable-next-line @next/next/no-img-element
      <img
        loading='eager'
        {...props}
        className={classNames(
          props.className,
          'max-w-full h-auto inline-block',
        )}
      />
    );
  }

  return (
    <Image
      loading='eager'
      {...props}
      priority={priority}
      className={classNames(props.className, 'max-w-full h-auto inline-block')}
    />
  );
};

export default CustomImage;
