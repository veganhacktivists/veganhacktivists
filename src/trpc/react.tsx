'use client';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { loggerLink, unstable_httpBatchStreamLink } from '@trpc/client';
import { createTRPCReact } from '@trpc/react-query';
import { type inferRouterInputs, type inferRouterOutputs } from '@trpc/server';
import { useState } from 'react';
import superjson from 'superjson';
import { Prisma } from '@prisma/client';

import type { AppRouter } from 'server/api/root';

const createQueryClient = () => new QueryClient();

let clientQueryClientSingleton: QueryClient | undefined = undefined;
const getQueryClient = () => {
  if (typeof window === 'undefined') {
    // Server: always make a new query client
    return createQueryClient();
  }
  // Browser: use singleton pattern to keep the same query client
  return (clientQueryClientSingleton ??= createQueryClient());
};

export const api = createTRPCReact<AppRouter>();

/**
 * Inference helper for inputs.
 *
 * @example type HelloInput = RouterInputs['example']['hello']
 */
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helper for outputs.
 *
 * @example type HelloOutput = RouterOutputs['example']['hello']
 */
export type RouterOutputs = inferRouterOutputs<AppRouter>;

superjson.registerCustom<Prisma.Decimal, string>(
  {
    isApplicable: (v): v is Prisma.Decimal => Prisma.Decimal.isDecimal(v),
    serialize: (v) => v.toJSON(),
    deserialize: (v) => new Prisma.Decimal(v),
  },
  'decimal.js',
);

export const TRPCReactProvider = (props: { children: React.ReactNode }) => {
  const queryClient = getQueryClient();

  const [trpcClient] = useState(() =>
    api.createClient({
      links: [
        loggerLink({
          enabled: (op) =>
            process.env.NODE_ENV === 'development' ||
            (op.direction === 'down' && op.result instanceof Error),
        }),
        unstable_httpBatchStreamLink({
          transformer: superjson,
          url: getBaseUrl() + '/api/trpc',
          headers: () => {
            return { 'x-trpc-source': 'nextjs-react' };
          },
        }),
      ],
    }),
  );

  return (
    <QueryClientProvider client={queryClient}>
      <api.Provider client={trpcClient} queryClient={queryClient}>
        {props.children}
      </api.Provider>
    </QueryClientProvider>
  );
};

function getBaseUrl() {
  if (typeof window !== 'undefined') return window.location.origin;
  if (process.env.URL) return `https://${process.env.URL}`;
  return `http://localhost:${process.env.PORT ?? 3000}`;
}
