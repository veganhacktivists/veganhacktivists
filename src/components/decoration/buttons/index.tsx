'use client';

import React, { useMemo } from 'react';
import {
  faInstagram,
  faPatreon,
  faYoutube,
} from '@fortawesome/free-brands-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classNames from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { faShare } from '@fortawesome/free-solid-svg-icons';
import { toast } from 'react-toastify';
import { FormattedMessage } from 'react-intl';

import useDeviceDetect from '../../../hooks/useDeviceDetect';

import { FillBackground } from './utils';

import type {
  MouseEventHandler,
  AnchorHTMLAttributes,
  Ref,
  HTMLAttributes,
} from 'react';
import type { LinkProps } from 'next/link';
import type ShareInfo from '../../layout/shareDialog/shareInfo';

export interface ButtonProps
  extends React.RefAttributes<HTMLElement>,
    HTMLAttributes<HTMLElement>,
    React.PropsWithChildren {
  primary?: boolean;
  href?: LinkProps['href'];
  className?: string;
  active?: boolean;
  linkProps?: Partial<LinkProps> & Partial<AnchorHTMLAttributes<unknown>>;
  onClick?: MouseEventHandler;
  type?: 'submit' | 'reset' | 'button';
  capitalize?: boolean;
  newTab?: boolean;
  disabled?: boolean;
}

const baseButtonClasses = classNames(
  'p-3 px-4 py-2 text-2xl border-l-8 transition-shadow font-mono cursor-pointer disabled:bg-grey-light disabled:cursor-not-allowed disabled:hover:shadow-none truncate focus:ring',
);

const isExternalLink: (href: ButtonProps['href']) => boolean = (href) => {
  if (!href) return false;
  if (typeof href === 'string') {
    return href.startsWith('http:') || href.startsWith('https:');
  }

  const { protocol, pathname, hostname } = href;

  return (
    protocol?.startsWith('mailto') ||
    [protocol, pathname, hostname].some(
      (x) => x?.startsWith('http') || x?.startsWith('https'),
    )
  );
};

const BaseButton = React.forwardRef<HTMLElement, ButtonProps>(
  (
    {
      children,
      linkProps,
      capitalize = true,
      className = '',
      newTab = false,
      ...props
    },
    ref,
  ) => {
    const classes = classNames(className, 'block', { capitalize });

    const isExternal = isExternalLink(props.href);

    const openInNewTab = useMemo(
      () => newTab || linkProps?.target || isExternal,
      [isExternal, linkProps?.target, newTab],
    );

    return (
      <>
        {props.href ? (
          <Link
            {...props}
            {...linkProps}
            href={props.href}
            passHref
            ref={ref as Ref<HTMLAnchorElement>}
            className={classNames(
              classes,
              props.disabled && 'bg-gray-light text-black',
            )}
            target={openInNewTab ? '_blank' : undefined}
            rel={openInNewTab ? 'noreferrer' : undefined}
          >
            {children}
          </Link>
        ) : (
          <button
            type='button'
            {...props}
            className={classes}
            ref={ref as Ref<HTMLButtonElement>}
          >
            {children}
          </button>
        )}
      </>
    );
  },
);

const LightButton: React.FC<ButtonProps> = ({
  children,
  primary,
  className = '',
  ...props
}) => {
  return (
    <FillBackground base='white' fill='green' disabled={props.disabled}>
      <BaseButton
        {...props}
        className={classNames(
          baseButtonClasses,
          'text-grey-dark border-green bg-w-x2 bg-white font-mono font-semibold',
          primary ? 'border-pink' : '',
          className,
        )}
      >
        {children}
      </BaseButton>
    </FillBackground>
  );
};

const DarkButton = React.forwardRef<HTMLElement, ButtonProps>(
  ({ children, active, className = '', ...props }, ref) => {
    return (
      <FillBackground
        disabled={props.disabled}
        base={active ? 'magenta' : 'grey-dark'}
        fill={active ? 'magenta' : 'green'}
      >
        <BaseButton
          {...props}
          ref={ref}
          className={classNames(
            baseButtonClasses,
            'transition-all overflow-hidden text-white',
            active ? 'border-pink' : 'border-green',
            className,
          )}
        >
          {children}
        </BaseButton>
      </FillBackground>
    );
  },
);

const DenyButton = React.forwardRef<HTMLElement, ButtonProps>(
  ({ children, active, className = '', ...props }, ref) => {
    return (
      <FillBackground
        disabled={props.disabled}
        base={active ? 'magenta' : 'grey-dark'}
        fill={active ? 'magenta' : '#FF1B18'}
      >
        <BaseButton
          {...props}
          ref={ref}
          className={classNames(
            baseButtonClasses,
            'transition-all overflow-hidden text-white',
            active ? 'border-pink' : 'border-[#FF1B18]',
            className,
          )}
        >
          {children}
        </BaseButton>
      </FillBackground>
    );
  },
);

const GreyButton: React.FC<ButtonProps> = ({
  className,
  children,
  ...props
}) => {
  return (
    <BaseButton
      {...props}
      className={classNames(
        baseButtonClasses,
        'bg-grey text-white border-none overflow-hidden',
        className,
      )}
    >
      {children}
    </BaseButton>
  );
};

const GreenButton: React.FC<ButtonProps> = ({
  children,
  primary,
  className = '',
  ...props
}) => {
  return (
    <FillBackground
      disabled={props.disabled}
      base={primary ? 'fuchsia' : 'green'}
      fill={primary ? 'pink' : 'green-dark'}
    >
      <BaseButton
        {...props}
        className={classNames(
          baseButtonClasses,
          'text-white border-green-dark',
          className,
        )}
      >
        {children}
      </BaseButton>
    </FillBackground>
  );
};

const BlueButton: React.FC<ButtonProps> = ({
  children,
  className = '',
  ...props
}) => {
  return (
    <FillBackground disabled={props.disabled} base={'blue'} fill={'blue-dark'}>
      <BaseButton
        {...props}
        className={classNames(
          baseButtonClasses,
          'text-white border-blue-dark',
          className,
        )}
      >
        {children}
      </BaseButton>
    </FillBackground>
  );
};

const ExternalLinkButton: React.FC<ButtonProps> = ({
  className,
  children,
  ...props
}) => {
  return (
    <FillBackground disabled={props.disabled} base='magenta' fill='pink-dark'>
      <BaseButton
        {...props}
        className={classNames(
          baseButtonClasses,
          className,
          'py-2 border-l-8 border-pink-dark',
        )}
      >
        {children}
      </BaseButton>
    </FillBackground>
  );
};

const IconButton: React.FC<ButtonProps> = (props) => {
  return <BaseButton {...props} />;
};

const PatreonButton: React.FC<ButtonProps> = ({ className, ...props }) => {
  return (
    <IconButton
      aria-label='Support us on Patreon!'
      {...props}
      className={classNames(
        'bg-white hover:bg-green-dark text-grey hover:text-white rounded-full px-2 py-2',
        className,
      )}
    >
      <div className='flex items-center justify-center w-8 h-8'>
        <FontAwesomeIcon size='2x' fixedWidth icon={faPatreon} />
      </div>
    </IconButton>
  );
};

const YoutubeButton: React.FC<ButtonProps> = ({ className, ...props }) => {
  return (
    <IconButton
      aria-label='Subscribe to our Youtube channel!'
      {...props}
      className={classNames(
        'bg-white hover:bg-red text-grey hover:text-white rounded-full px-2 py-2',
        className,
      )}
    >
      <div className='flex items-center justify-center w-8 h-8'>
        <FontAwesomeIcon size='2x' fixedWidth icon={faYoutube} />
      </div>
    </IconButton>
  );
};

const InstagramButton: React.FC<ButtonProps> = ({ className, ...props }) => {
  return (
    <IconButton
      aria-label='Follow us on Instagram!'
      {...props}
      className={classNames(
        'bg-white hover:bg-blue text-grey hover:text-white rounded-full px-2 py-2',
        className,
      )}
    >
      <div className='flex items-center justify-center w-8 h-8'>
        <FontAwesomeIcon size='2x' fixedWidth icon={faInstagram} />
      </div>
    </IconButton>
  );
};

const ShareButton: React.FC<
  ButtonProps & {
    openAndInitiateShareDialog: () => void;
    shareInfo: ShareInfo;
  }
> = ({ openAndInitiateShareDialog, shareInfo }) => {
  const { isMobile, isReady } = useDeviceDetect();
  const { url, title, description } = shareInfo;

  const shareNatively = async () => {
    await navigator.share({ title, text: description, url });
  };

  const shareWithDialog = () => {
    openAndInitiateShareDialog();
  };

  return (
    <DarkButton
      disabled={!isReady}
      onClick={() => {
        if (navigator && navigator['share'] && isMobile) {
          void toast.promise(shareNatively, {
            error:
              'Something went wrong while sharing your content! Please try again later',
          });
        } else {
          shareWithDialog();
        }
      }}
      linkProps={{ scroll: false }}
    >
      <div className='flex items-center justify-center gap-2'>
        <FontAwesomeIcon size='1x' fixedWidth icon={faShare} />
        <div>
          <FormattedMessage
            id='component.button.share.label'
            defaultMessage='Share'
          />
        </div>
      </div>
    </DarkButton>
  );
};

const WhiteButton: React.FC<ButtonProps> = ({
  children,
  active,
  className = '',
  ...props
}) => {
  const classes = classNames(
    'border-2 border-opacity-50 p-3',
    {
      'bg-gray text-white border-grey': active,
      'text-black': !active,
    },
    className,
  );

  return (
    <FillBackground base='white' fill='green' disabled={props.disabled}>
      <BaseButton {...props}>
        <div className={classes}>{children}</div>
      </BaseButton>
    </FillBackground>
  );
};

const NavButton: React.FC<ButtonProps & { href: string }> = ({
  href,
  className = '',
  children,
}) => {
  const { pathname } = useRouter();
  const atLocation = pathname === href;

  const classes = classNames('m-5 font-mono', className);

  return (
    <DarkButton
      active={atLocation}
      href={href}
      className={classes}
      linkProps={{ scroll: false }}
    >
      {children}
    </DarkButton>
  );
};

const OutlineButton: React.FC<ButtonProps> = ({
  className,
  active = false,
  ...props
}) => {
  return (
    <BaseButton
      {...props}
      className={classNames(
        baseButtonClasses,
        className,
        'border-grey-dark border-l border focus:ring-0 bg-none',
        {
          'text-grey-dark': !active,
          'text-white bg-grey-dark': active,
        },
      )}
    />
  );
};

const LogoutButton: React.FC<ButtonProps> = ({
  children,
  primary,
  className = '',
  ...props
}) => {
  return (
    <FillBackground base='white' fill='#FF1A18' disabled={props.disabled}>
      <BaseButton
        {...props}
        className={classNames(
          baseButtonClasses,
          'text-grey-dark border-[#FF1A18] bg-w-x2 bg-white font-mono font-semibold',
          primary ? 'border-pink' : '',
          className,
        )}
      >
        {children}
      </BaseButton>
    </FillBackground>
  );
};

export {
  ExternalLinkButton,
  PatreonButton,
  InstagramButton,
  IconButton,
  YoutubeButton,
  ShareButton,
  LightButton,
  DarkButton,
  WhiteButton,
  GreenButton,
  NavButton,
  GreyButton,
  OutlineButton,
  DenyButton,
  LogoutButton,
  BlueButton,
};
