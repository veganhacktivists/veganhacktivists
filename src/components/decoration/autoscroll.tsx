'use client';

import classNames from 'classnames';
import { useMemo } from 'react';

export interface AutoScrollProps {
  items: React.ReactNode[] | React.ReactNode[][];
}

interface AnimatedRowProps {
  items: React.ReactNode[];
  backwards?: boolean;
}

const AnimatedRow = ({ items, backwards }: AnimatedRowProps) => {
  return (
    <div className='overflow-hidden'>
      <div
        className={classNames(
          'flex flex-row w-fit gap-8 place-items-center',
          backwards
            ? 'animate-infinite-scroll-reverse'
            : 'animate-infinite-scroll',
        )}
      >
        {[...items, ...items].map((item, i) => (
          <li key={i} className='w-max max-w-[15rem] flex-shrink-0 last:mr-8'>
            <div className='grayscale hover:grayscale-0 transition-[filter] ease-out'>
              {item}
            </div>
          </li>
        ))}
      </div>
    </div>
  );
};

export const Autoscroll = ({ items }: AutoScrollProps) => {
  const rows = useMemo(
    () => (Array.isArray(items[0]) ? items : [items]) as React.ReactNode[][],
    [items],
  );

  return (
    <ul className='space-y-12'>
      {rows.map((row, i) => (
        <AnimatedRow key={i} items={row} backwards={i % 2 === 0} />
      ))}
    </ul>
  );
};
