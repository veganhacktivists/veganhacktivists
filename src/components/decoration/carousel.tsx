'use client';

import classNames from 'classnames';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import React from 'react';

import useWindowSize from 'hooks/useWindowSize';

import type { UIEventHandler } from 'react';

export interface CarouselProps {
  items: React.ReactNode[];
  layout?: 'horizontal' | 'grid';
  theme?: 'light' | 'dark';
  pageWidth?: number;
  itemWidth?: number;
}

export const Carousel = ({
  items,
  layout = 'horizontal',
  theme = 'light',
  pageWidth = 3,
  itemWidth = 256,
}: CarouselProps) => {
  const [currentPage, setCurrentPage] = useState(0);

  const { width = 1920 } = useWindowSize();

  const numberOfRows = layout === 'grid' ? pageWidth : 1;

  const itemsPerPage =
    Math.min(Math.max(Math.floor(width / itemWidth), 1), pageWidth) *
    numberOfRows;

  const numPages = Math.ceil(items.length / itemsPerPage);

  const listItemsRef = useRef<(HTMLLIElement | null)[]>([]);

  useEffect(() => {
    listItemsRef.current = listItemsRef.current.slice(0, items.length);
  }, [items.length]);

  const pageBreakpoints = useMemo(() => {
    const pageWidth = (itemWidth + 16) * itemsPerPage;
    return [...new Array(numPages)].map((_, i) => {
      return pageWidth * i;
    });
  }, [itemWidth, itemsPerPage, numPages]);

  const handleScroll = useCallback<UIEventHandler<HTMLUListElement>>(
    (e) => {
      const target = e.target as HTMLUListElement;
      const scrollLeft = target.scrollLeft;

      const page =
        pageBreakpoints.findIndex((point) => scrollLeft <= point) ??
        numPages - 1;

      setCurrentPage(page);
    },
    [numPages, pageBreakpoints],
  );

  const getHandlePageChange = useCallback(
    (newPage: number) => () => {
      const itemIndex = newPage * itemsPerPage;
      const item = listItemsRef.current[itemIndex];

      item?.scrollIntoView({
        block: 'nearest',
        inline: 'start',
      });
    },
    [itemsPerPage],
  );

  return (
    <div
      className='space-y-16 mx-auto w-full'
      style={{
        // (16rem w + 1rem gap) * items x - 1rem gap
        maxWidth: `${
          (itemWidth / 16 + 1) * Math.ceil(itemsPerPage / numberOfRows) - 1
        }rem`,
      }}
    >
      <ul
        onScroll={handleScroll}
        className={classNames(
          'overflow-auto scrollbar-none snap-x snap-mandatory gap-4 scroll-smooth',
          {
            'flex flex-row flex-nowrap': layout === 'horizontal',
            'grid grid-rows-3 grid-flow-col': layout === 'grid',
          },
        )}
      >
        {items.map((item, i) => (
          <li
            key={i}
            className={classNames('flex-shrink-0', {
              'snap-start': i % itemsPerPage === 0,
            })}
            style={{
              width: itemWidth,
            }}
            ref={(ref) => {
              listItemsRef.current[i] = ref;
            }}
          >
            {item}
          </li>
        ))}
      </ul>
      <div className='flex flex-row gap-4 justify-center items-center flex-wrap'>
        {[...Array(numPages)].map((_, i) => (
          <button
            aria-label={`Scroll to page ${i + 1}`}
            onClick={getHandlePageChange(i)}
            type='button'
            key={i}
            className={classNames(
              'w-4 h-4 flex-shrink-0',
              currentPage === i
                ? theme === 'light'
                  ? 'bg-white'
                  : 'bg-gray-dark'
                : 'bg-[#737373]',
            )}
          />
        ))}
      </div>
    </div>
  );
};
