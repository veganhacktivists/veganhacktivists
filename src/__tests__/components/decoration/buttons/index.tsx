import React from 'react';
import { render } from '@testing-library/react';
import { WhiteButton } from '../../../../components/decoration/buttons';

describe('WhiteButton', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <WhiteButton href="https://example.org">Foo</WhiteButton>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
