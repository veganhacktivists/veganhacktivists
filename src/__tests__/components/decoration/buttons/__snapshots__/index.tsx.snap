// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhiteButton should render correctly 1`] = `
<DocumentFragment>
  <a
    href="https://example.org"
    rel="noreferrer"
    target="_blank"
  >
    <div
      class="border-2 border-gray-200 border-opacity-50 p-3"
    >
      Foo
    </div>
  </a>
</DocumentFragment>
`;
